core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+
+    private Builder() {}
+
+    private Builder(ViewMetadata base) {
+      this.formatVersion = base.formatVersion();
+      this.location = base.location();
+      this.currentSchemaId = base.currentSchemaId();
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder formatVersion(int newFormatVersion) {
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder location(String newLocation) {
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder currentVersionId(int versionId) {
+      this.currentVersionId = versionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(versionId));
+      return this;
+    }
+
+    public Builder currentSchemaId(Integer schemaId) {
+      this.currentSchemaId = schemaId;
+      this.changes.add(new MetadataUpdate.SetCurrentSchema(schemaId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder addAllSchemas(Iterable<Schema> schemaEntries) {
+      for (Schema schema : schemaEntries) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      }
+      return this;
+    }
+
+    public Builder schemas(Iterable<Schema> newSchemas) {
+      this.schemas = Lists.newArrayList(newSchemas);
+      for (Schema schema : newSchemas) {
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      }
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      return this;
+    }
+
+    public Builder addAllVersions(Iterable<ViewVersion> versionEntries) {
+      for (ViewVersion version : versionEntries) {
+        this.versions.add(version);
+        this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      }
+      return this;
+    }
+
+    public Builder versions(Iterable<ViewVersion> newVersions) {
+      this.versions = Lists.newArrayList(newVersions);
+      for (ViewVersion version : newVersions) {
+        this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      }
+      return this;
+    }
+
+    public Builder addHistory(ViewHistoryEntry historyEntry) {
```

Comment:
History should be maintained internally. We don't expect the caller to provide this.

Reply:
fixed now